src/impl/apis/data_api_impl.py

```diff
@@ -258,35 +261,167 @@ async def get_time_range(
+    if is_single_value:
+        spatial_unit_id_index = column_names.index("spatial_unit_id")
+    is_flow = "origin" in column_names and "destination" in column_names
+    if is_flow:
+        origin_index = column_names.index("origin")
+        destination_index = column_names.index("destination")
+    data_by_date = {}
+    min_value = math.inf
+    max_value = -math.inf
+    num_rows = 0
+    async for chunk in stream_query(base_table_name, metadata, pool, table_name):
+        if chunk is None:
+            # I don't think I should need to do this?
+            break
+        for row in chunk:
+            num_rows += 1
+            # adjust the global min/max if necessary
+            min_value = min([row[data_index], min_value])
+            max_value = max([row[data_index], max_value])
+            this_date = mdid_to_date(row[mdid_index], metadata).strftime(tr.date_format)
+            data_by_date.setdefault(this_date, {})
+            value = util.num(str(row[data_index]))
+            if is_single_value:
+                data_by_date[this_date][row[spatial_unit_id_index]] = value
+            elif is_flow:
+                data_by_date[this_date].setdefault(row[origin_index], {})
+                if value is not None:
+                    data_by_date[this_date][row[origin_index]][row[destination_index]] = value
+            else:
+                raise HTTPException(
+                    status_code=HTTPStatus.INTERNAL_SERVER_ERROR,
+                    detail="Cannot find spatial unit in data",
+                )
+
+    logger.debug(
+        "Finished running data query",
+        num_results=num_rows,
+        min_value=min_value,
+        max_value=max_value,
+    )
+    if len(data_by_date.keys()) == 0:
+        raise HTTPException(status_code=HTTPStatus.NOT_FOUND, detail="No data found")
+
+    new_result = QueryResult(
+        min=min_value if min_value != math.inf else None,
+        max=max_value if max_value != -math.inf else None,
+        data_by_date=dict(sorted(data_by_date.items())),
+    )
+
+    return new_result
+
+
+async def get_column_names(table_name: str, metadata: List[Metadata], pool: Pool) -> List:
+    select_query = (
+        f"SELECT * FROM `{os.getenv('DB_NAME')}`.`{table_name}_{metadata[0].mdid}` LIMIT 1;"
+    )
+    logger.debug("Getting column names", table_name=table_name)
+    async with pool.acquire() as conn, conn.cursor() as cursor:
+        await cursor.execute(select_query)
+        return [i[0] for i in cursor.description]
+
+
+async def stream_query(
+    base_table_name: str, metadata: List[Metadata], pool: Pool, table_name: str
+) -> AsyncGenerator[list, None]:
+    table_names = [f"`{os.getenv('DB_NAME')}`.`{table_name}_{m.mdid}`" for m in metadata]
+    short_names = [table_name.rpartition(".")[2].strip("`") for table_name in table_names]
+
+    partition_queries = [
+        f"SELECT `{short_name}`.*, `dt` FROM {table_name} AS `{short_name}` LEFT JOIN metadata USING (mdid)"
```

Comment:
We actually know the column names at this point, right? So it saves us some effort later to get them explicitly here, and avoid returning an extra column we don't want to return.

Reply:
Not sure I follow what you mean here?

Reply:
Oh, I see - explicit columns instead of the `*`

Reply:
I think the dict cursor largely obviates it really, but we're getting the mdid column here which we then don't use.